packages/api/src/useSubscription.ts

```diff
@@ -1,42 +1,62 @@
-import { useState } from 'react';
-import { useWS as useWSShared } from '@deriv/shared';
-import { TSocketSubscribableEndpointNames, TSocketAcceptableProps, TSocketResponseData } from '../types';
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { useWS } from '@deriv/shared';
+import type {
+    TSocketSubscribableEndpointNames,
+    TSocketAcceptableProps,
+    TSocketResponseData,
+    TSocketRequestCleaned,
+} from '../types';
 
 const useSubscription = <T extends TSocketSubscribableEndpointNames>(name: T) => {
     const [is_loading, setIsLoading] = useState(false);
     const [is_subscribed, setSubscribed] = useState(false);
     const [error, setError] = useState<unknown>();
     const [data, setData] = useState<TSocketResponseData<T>>();
-    const [subscriber, setSubscriber] = useState<{ unsubscribe?: VoidFunction }>();
-    const WS = useWSShared();
-
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const onData = (response: any) => {
-        setData(response[name === 'ticks' ? 'tick' : name]);
-        setIsLoading(false);
-    };
-
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const onError = (response: any) => {
-        setError(response.error);
-        setIsLoading(false);
-    };
-
-    const subscribe = (...props: TSocketAcceptableProps<T>) => {
-        setIsLoading(true);
-        setSubscribed(true);
-
-        try {
-            setSubscriber(WS.subscribe({ [name]: 1, subscribe: 1, ...(props[0] || {}) }).subscribe(onData, onError));
-        } catch (e) {
-            setError(e);
-        }
-    };
-
-    const unsubscribe = () => {
-        subscriber?.unsubscribe?.();
+    const subscriber = useRef<{ unsubscribe?: VoidFunction }>();
+    const WS = useWS();
+
+    const subscribe = useCallback(
+        (...props: TSocketAcceptableProps<T>) => {
+            const prop = props?.[0];
+            const payload = prop && 'payload' in prop ? (prop.payload as TSocketRequestCleaned<T>) : undefined;
+
+            setIsLoading(true);
+            setSubscribed(true);
+
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            const onData = (response: any) => {
```

Comment:
```suggestion
            const onData = (response: any) => {
```
Can we type using the generics that you declared? `TSocketResponseData<T>` or something

Reply:
@yashim-deriv This is at the library level, We are typing the `any` here and exposing typesafe data to the outside.